### PR TITLE
[SPARK-44959][BUILD] Upgrade sbt to 1.9.5

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to upgrade sbt from 1.9.3 to 1.9.4.

### Why are the changes needed?
1.The 1.9.4 fix a security vulnerability: 
[CVE-2022-46751](https://github.com/advisories/GHSA-2jc4-r94c-rp7h) is a security vulnerability discovered in Apache Ivy, but found also in Coursier.
image

2.Full release notes:
https://github.com/sbt/sbt/releases/tag/v1.9.4

3.v1.9.3 VS v1.9.4
https://github.com/sbt/sbt/compare/v1.9.3...v1.9.4

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by dongjoon-hyun — DISMISSED** [review]
+1, LGTM (Pending CIs). Thank you, @panbingkun .

cc @LuciferYang

**Review by dongjoon-hyun — COMMENTED** [review]
Scala 2.13 test seems to fail. Could you check it locally, @panbingkun ?

Screenshot 2023-08-25 at 1 35 06 AM

**Comment by panbingkun** [comment]
> Scala 2.13 test seems to fail. Could you check it locally, @panbingkun ?
> 
> Screenshot 2023-08-25 at 1 35 06 AM

Please do not merge in until I confirm. I will carefully check and it seems to be a cache issue at first glance.

**Comment by LuciferYang** [comment]
no problem with the local build, I think you can try again. @panbingkun

**Comment by panbingkun** [comment]
> no problem with the local build, I think you can try again. @panbingkun

But I also had a similar issue with the local run, let me investigate it first.
fc5abb533c56415876c10eb72760520f

**Comment by panbingkun** [comment]
It should be a cache issue with the Courier. When I delete `4.1.96.Final` from the local mvn report and execute it again, the error disappears.

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -237,7 +237,7 @@ jobs:
       uses: actions/cache@v3
       with:
         path: ~/.cache/coursier
-        key: ${{ matrix.java }}-${{ matrix.hadoop }}-coursier-${{ hashFiles('**/pom.xml', '**/plugins.sbt') }}
```
Comment (panbingkun) [inline .github/workflows/build_and_test.yml:240]: only for test.

**Comment by dongjoon-hyun** [comment]
Oh, IIUC, although we cleared the cache, the last commit seems to show `Scala 2.13 SBT` failure still.

**Comment by panbingkun** [comment]
Yeah, Let me continue investigating it.

**Comment by dongjoon-hyun** [comment]
Thank you for retry. It seems that Scala 2.13 SBT still fails with `useCoursier := false`.

**Comment by LuciferYang** [comment]
Can we try to delete `~/.cache/coursier/v1/https` as well?

File: .github/workflows/build_and_test.yml [hunk 2]
```diff
@@ -856,7 +864,7 @@ jobs:
     - name: Build with SBT
       run: |
         ./dev/change-scala-version.sh 2.13
-        ./build/sbt -Pyarn -Pmesos -Pkubernetes -Pvolcano -Phive -Phive-thriftserver -Phadoop-cloud -Pkinesis-asl -Pdocker-integration-tests -Pkubernetes-integration-tests -Pspark-ganglia-lgpl -Pscala-2.13 compile Test/compile
+        ./build/sbt -Pyarn -Pmesos -Pkubernetes -Pvolcano -Phive -Phive-thriftserver -Phadoop-cloud -Pkinesis-asl -Pdocker-integration-tests -Pkubernetes-integration-tests -Pspark-ganglia-lgpl -Pscala-2.13 compile Test/compile/ useCoursier := false
```
Comment (LuciferYang) [inline .github/workflows/build_and_test.yml:867]: should be `-Dsbt.coursier=false`.

Comment (LuciferYang) [reply .github/workflows/build_and_test.yml:867]: And `useCoursier := false` should be configured in the `build.sbt` file rather than in the command line.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Although the CI passed, this PR seems to be not ready because it has several mixed themes in a single PR.

- The change `build_and_test.yml` should be in an independent PR, if that is required.
- `plugins.sbt` should be in an independent PR, if that is required.

**Comment by dongjoon-hyun** [comment]
Gentle ping, @panbingkun . How do you want to proceed this?

**Comment by panbingkun** [comment]
> Gentle ping, @panbingkun . How do you want to proceed this?

@dongjoon-hyun 
Let me think carefully, We may need to put workflow `scala-213` into the container run first, which is safer to delete some information, eg: `rm -rf /home/runner/.m2`

**Comment by LuciferYang** [comment]
@panbingkun sbt 1.9.5 released: https://github.com/sbt/sbt/releases/tag/v1.9.5

Can you reuse this ticket to verify it?

**Comment by panbingkun** [comment]
> @panbingkun sbt 1.9.5 released: https://github.com/sbt/sbt/releases/tag/v1.9.5
> 
> Can you reuse this ticket to verify it?

Okay, let me to verify it.

**Comment by LuciferYang** [comment]
could you rebase this one ? @panbingkun

**Comment by panbingkun** [comment]
> could you rebase this one ? @panbingkun
I have submitted a test on the latest master, but unfortunately it still has issues. I have identified the issue, but I need some time to see how to fix it reasonably.
https://github.com/panbingkun/spark/actions/runs/6294142093/job/17087611901

**Comment by LuciferYang** [comment]
> > could you rebase this one ? @panbingkun
> > I have submitted a test on the latest master, but unfortunately it still has issues. I have identified the issue, but I need some time to see how to fix it reasonably.
> > https://github.com/panbingkun/spark/actions/runs/6294142093/job/17087611901

It appears that only lint-java downloads some jar files? Can we operate without using Maven cache? What would be the impact on performance?

**Comment by panbingkun** [comment]
> https://github.com/panbingkun/spark/actions/runs/6294142093/job/17087611901

Its root cause is that if there is a corrupted jar in Maven's repo, the above problem will occur. Currently, the possible solutions to bypass it are
1.Before run, delete all Maven's repos
2.Adjust the cache policy of the courier

Yes, it all affects run time to some extent. Currently, what I am looking for is a solution that does not affect run time, give me some time.

**Comment by LuciferYang** [comment]
```
1.Before run, delete all Maven's repos
```
Does this equate to removing the `Cache Maven local repository` step in lint?

**Comment by LuciferYang** [comment]
> Its root cause is that if there is a corrupted jar in Maven's repo

I'm curious about this conclusion, why didn't the use of Maven to build `java-other-versions` trigger similar issues?

**Pull request closed without merging**